OpenRA.Mods.Common/Traits/World/ClusterManager.cs

```diff
@@ -0,0 +1,314 @@
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	abstract class ClusterManager
+	{
+		protected readonly CellLayer<ClusterContents> clusterCellLayer;
+
+		protected readonly ClusterContents emptyCluster;
+		protected readonly List<ClusterContents> clusters = new List<ClusterContents>();
+		protected readonly World world;
+		protected readonly Map map;
+
+		public ClusterManager(World world)
+		{
+			this.world = world;
+			map = world.Map;
+			clusterCellLayer = new CellLayer<ClusterContents>(world.Map);
+			emptyCluster = new ClusterContents(true);
+			clusters.Add(emptyCluster);
+		}
+
+		protected abstract bool ClusterCondition(CPos p);
+
+		public bool IsConnected(CPos p1, CPos p2)
+		{
+			if (!clusterCellLayer.Contains(p1) || !clusterCellLayer.Contains(p2))
+				return false;
+
+			if (clusterCellLayer[p1] == clusterCellLayer[p2])
+				return true;
+
+			// Even though p1 and p2 are in different domains, it's possible
+			// that some dynamic terrain (i.e. bridges) may connect them.
+			return clusterCellLayer[p1].TransientConnections.Any(con => con.Active && clusterCellLayer[p2].TransientConnections.Contains(con));
+		}
+
+		public void UpdateCells(HashSet<CPos> dirtyCells)
+		{
+			var addedCells = dirtyCells.Where(cell => ClusterCondition(cell) && clusterCellLayer[cell].IsEmpty);
+			var removedCells = dirtyCells.Where(cell => !ClusterCondition(cell) && !clusterCellLayer[cell].IsEmpty);
+
+			foreach (var cell in addedCells)
+			{
+				var neighborClusters = new List<ClusterContents>();
+
+				// Select all neighbor clusters
+				var neighbors = CVec.Directions.Select(d => d + cell)
+					.Where(c => map.Contains(c) && !removedCells.Contains(c) && ClusterCondition(c));
+
+				ClusterContents foundCluster = null;
+				foreach (var n in neighbors)
+				{
+					if (foundCluster == null)
+					{
+						// Join the first viable neighbor we find.
+						foundCluster = clusterCellLayer[n];
+						UpdateCell(cell, clusterCellLayer[n]);
+					}
+					else
+					{
+						// Assimilate nearby neighbors
+						AssignCluster(clusterCellLayer[n], foundCluster);
+					}
+				}
+
+				if (foundCluster == null)
+				{
+					// If we can't join a neighbor, create a new cluster
+					UpdateCell(cell, new ClusterContents());
+					clusters.Add(clusterCellLayer[cell]);
+				}
+			}
+
+			foreach (var cell in removedCells)
+			{
+				// We don't need to test for splitting if it's the last cell in a domain
+				if (clusterCellLayer[cell].Cells.Count > 1)
+				{
+					// Select all neighbors inside the map boundaries
+					var neighbors = CVec.Directions.Select(d => d + cell)
+						.Where(c => clusterCellLayer[cell].Cells.Contains(c) && !removedCells.Contains(c)).ToList();
+
+					// We don't need to test for splitting if it's an edge of a domain
```

Comment:
This needs to be a smarter check, we don't want to be recalculating the whole cluster when it is clearly not needed

Reply:
IIrc that was the idea behind additionally dividing the map into chunks.

Reply:
Destroying a bridge on the Hypothermia map in Red Alert freezes the game for ~3 seconds whilst it does this recalculation. The current domain index avoids the performance issue by being incorrect and not updating at all.

Even if we divided the area up into grids, we still suffer a large costs because we are trying to maintain an up-to-date list of cells within each domain. When you need to split a large domain into two, there's no way to get around the fact you're going to have to update some large lists.

The hierarchical path finder divides the map up into grids, and tracks connectivity at that level. If a split occurs, it only has to rebuild the connections on that grid. It gets away with that because it is *not* trying to keep a specific list of cells in each domain index. That means if you want to know if cells belong to the same domain index, you still need to invoke a path search to see if they are connected! The "hierarchical" part means this is reasonably cheap because you can consult the much smaller abstract graph to determine connected regions, rather than having to consult the full sized map.

The hierarchical path finder would allow us to replace queries against the domain index that are related to pathing, e.g. when the naval AI is trying to find the nearest enemy to attack that is reachable over water. For other considerations like this PR was working towards such as tracking clusters of ore and gems, we'd need to separate out the "hierarchical" logic from the "pathfinding" logic in order to reuse it.

Coming back to focus on this PR on its own: I'm wondering if making the domain index correct by having it handle more updates is worth the performance cost that incurs in the short term. Given we've gotten away with it for years maybe the lesser evil is to keep it as is, much as that pains me to suggest as this is a nice PR conceptually.

Reply:
The recalculation is not yet optimised well, it has a few glaring problems that I'll address hopefully soon. When testing I haven't experienced any lag-spikes but I guess I now have a good testcase.